### PR TITLE
HTML API: Implement active format reconstruction

### PR DESCRIPTION
**Pull request by dmsnell**
Trac ticket: Core-61576

Adds support for _active format reconstruction_, which occurs when crossing certain HTML boundaries, such as when entering a new `P` element which implicitly closed the previous one and all of the formatting elements inside it.

This *raises the question* what to do when elements are implicitly created. This appears already with the unexpected `</p>`, which creates an empty `P` element. `next_tag()` never finds these elements even though they appear in the breadcrumbs _when moving past them_.

```html
<p><b>This is bold<p>This is a new paragraph, and it's still bold.
```

| [DOM in browser](https://software.hixie.ch/utilities/js/live-dom-viewer/) | HTML Breadcrumbs |
|---|---|
| Screenshot 2024-01-13 at 9 29 32 AM | Screenshot 2024-01-13 at 9 30 03 AM |

### CODE REVIEW

File: src/wp-includes/html-api/class-wp-html-active-formatting-elements.php [hunk 1]
```diff
@@ -43,6 +43,20 @@ class WP_HTML_Active_Formatting_Elements {
...
+	 * Returns the node at the given index in the list of active formatting elements.
+	 *
+	 * Do not use this method; it is meant to be used only by the HTML Processor.
+	 *
+	 * @access private
+	 *
+	 * @param int $index Number of nodes from the top node to return.
+	 * @return WP_HTML_Token|null Node at the given index in the stack, if one exists, otherwise null.
+	 */
+	public function at( $index ) {
```
Comment (sirreal) [inline src/wp-includes/html-api/class-wp-html-active-formatting-elements.php:58]: Maybe some private looking prefix is a good idea to suggest "this isn't something that should be used publicly":

```suggestion
	public function _at( $index ) {
```

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 2]
```diff
@@ -1472,15 +1472,19 @@ private function generate_implied_end_tags_thoroughly() {
 	 * @return bool Whether any formatting elements needed to be reconstructed.
```
Comment (sirreal) [inline src/wp-includes/html-api/class-wp-html-processor.php:2832]: I wonder if `false | int` would be helpful, where we could return `false` or the number of reconstructed elements.

File: src/wp-includes/html-api/class-wp-html-active-formatting-elements.php (hunk 1 above)
Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-active-formatting-elements.php:58]: this is what `@access private` is for. we shouldn't need to apply any other mechanisms.

File: src/wp-includes/html-api/class-wp-html-processor.php (hunk 2 above)
Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-processor.php:2832]: can you think of a case where it would be useful?

**Comment by dmsnell** [comment]
Superseded by https://github.com/WordPress/wordpress-develop/pull/6982

**Pull request closed without merging**